File: tech-studio-projects/financial-industry/algo-trading-platform/cpp/orderbook/Orderbook.cpp
#include "Orderbook.h"

// Compare functor for ordering Ask Limits.
bool AskLimitComparer::operator()(const std::shared_ptr<Limit>& lhs, const std::shared_ptr<Limit>& rhs) const {
    return lhs->GetPrice() < rhs->GetPrice(); // Replace with actual comparison logic.
}

// Compare functor for ordering Bid Limits.
bool BidLimitComparer::operator()(const std::shared_ptr<Limit>& lhs, const std::shared_ptr<Limit>& rhs) const {
    return lhs->GetPrice() > rhs->GetPrice(); // Replace with actual comparison logic.
}

// Constructor implementation.
Orderbook::Orderbook(std::shared_ptr<Security> instrument)
    : instrument(instrument) {
}

// Getter for the number of orders.
int Orderbook::GetCount() const {
    return orders.size();
}

// Add an order to the order book.
void Orderbook::AddOrder(const std::shared_ptr<Order>& order) {
    auto baseLimit = std::make_shared<Limit>(order->GetPrice());
    AddOrderToLimit(order, baseLimit, order->IsBuySide() ? bidLimits : askLimits, orders);
}

// Static method to add an order to a limit level.
void Orderbook::AddOrderToLimit(const std::shared_ptr<Order>& order,
    const std::shared_ptr<Limit>& baseLimit,
    std::set<std::shared_ptr<Limit>, BidLimitComparer>& limitLevels,
    std::map<long, std::shared_ptr<OrderbookEntry>>& internalBook) {
    auto it = limitLevels.find(baseLimit);
    if (it != limitLevels.end()) {
        auto limit = *it;
        auto orderbookEntry = std::make_shared<OrderbookEntry>(order, baseLimit);
        if (limit->GetHead() == nullptr) {
            limit->SetHead(orderbookEntry);
            limit->SetTail(orderbookEntry);
        }
        else {
            auto tailPointer = limit->GetTail();
            tailPointer->SetNext(orderbookEntry);
            orderbookEntry->SetPrevious(tailPointer);
            limit->SetTail(orderbookEntry);
        }
        internalBook[order->GetOrderId()] = orderbookEntry;
    }
    else {
        limitLevels.insert(baseLimit);
        auto orderbookEntry = std::make_shared<OrderbookEntry>(order, baseLimit);
        baseLimit->SetHead(orderbookEntry);
        baseLimit->SetTail(orderbookEntry);
        internalBook[order->GetOrderId()] = orderbookEntry;
    }
}

// Change an existing order.
void Orderbook::ChangeOrder(const std::shared_ptr<ModifyOrder>& modifyOrder) {
    auto it = orders.find(modifyOrder->GetOrderId());
    if (it != orders.end()) {
        auto orderbookEntry = it->second;
        RemoveOrderFromLimit(modifyOrder->ToCancelOrder(), orderbookEntry, orders);
        AddOrder(modifyOrder->ToNewOrder());
    }
}

// Check if an order exists.
bool Orderbook::ContainsOrder(long orderId) const {
    return orders.find(orderId) != orders.end();
}

// Get all bid orders.
std::list<std::shared_ptr<OrderbookEntry>> Orderbook::GetBidOrders() const {
    std::list<std::shared_ptr<OrderbookEntry>> orderbookEntries;
    for (const auto& bidLimit : bidLimits) {
        if (bidLimit->IsEmpty()) continue;
        auto entry = bidLimit->GetHead();
        while (entry != nullptr) {
            orderbookEntries.push_back(entry);
            entry = entry->GetNext();
        }
    }
    return orderbookEntries;
}

// Get all ask orders.
std::list<std::shared_ptr<OrderbookEntry>> Orderbook::GetAskOrders() const {
    std::list<std::shared_ptr<OrderbookEntry>> orderbookEntries;
    for (const auto& askLimit : askLimits) {
        if (askLimit->IsEmpty()) continue;
        auto entry = askLimit->GetHead();
        while (entry != nullptr) {
            orderbookEntries.push_back(entry);
            entry = entry->GetNext();
        }
    }
    return orderbookEntries;
}

// Get the current spread.
OrderbookSpread Orderbook::GetSpread() const {
    std::shared_ptr<Limit> bestAsk = nullptr;
    std::shared_ptr<Limit> bestBid = nullptr;

    if (!askLimits.empty() && !askLimits.begin()->IsEmpty()) {
        bestAsk = *askLimits.begin();
    }

    if (!bidLimits.empty() && !bidLimits.rbegin()->IsEmpty()) {
        bestBid = *bidLimits.rbegin();
    }

    return OrderbookSpread(bestBid ? std::make_optional(bestBid->GetPrice()) : std::nullopt,
        bestAsk ? std::make_optional(bestAsk->GetPrice()) : std::nullopt);
}

// Remove an order from the order book.
void Orderbook::RemoveOrder(const std::shared_ptr<CancelOrder>& cancelOrder) {
    auto it = orders.find(cancelOrder->GetOrderId());
    if (it != orders.end()) {
        RemoveOrderFromLimit(cancelOrder, it->second, orders);
    }
}

// Static method to remove an order from a limit level.
void Orderbook::RemoveOrderFromLimit(const std::shared_ptr<CancelOrder>& cancelOrder,
    std::shared_ptr<OrderbookEntry> orderbookEntry,
    std::map<long, std::shared_ptr<OrderbookEntry>>& internalBook) {
    auto limit = orderbookEntry->GetParentLimit();
    if (orderbookEntry->GetPrevious() && orderbookEntry->GetNext()) {
        orderbookEntry->GetNext()->SetPrevious(orderbookEntry->GetPrevious());
        orderbookEntry->GetPrevious()->SetNext(orderbookEntry->GetNext());
    }
    else if (orderbookEntry->GetPrevious()) {
        orderbookEntry->GetPrevious()->SetNext(nullptr);
    }
    else if (orderbookEntry->GetNext()) {
        orderbookEntry->GetNext()->SetPrevious(nullptr);
    }

    if (limit->GetHead() == orderbookEntry && limit->GetTail() == orderbookEntry) {
        limit->SetHead(nullptr);
        limit->SetTail(nullptr);
    }
    else if (limit->GetHead() == orderbookEntry) {
        limit->SetHead(orderbookEntry->GetNext());
    }
    else if (limit->GetTail() == orderbookEntry) {
        limit->SetTail(orderbookEntry->GetPrevious());
    }

    internalBook.erase(cancelOrder->GetOrderId());
}

File: tech-studio-projects/financial-industry/algo-trading-platform/java/algo-trading-platform/src/com/fintech/algotrading/orderbook/Orderbook.java
package com.fintech.algotrading.orderbook;

import java.util.*;

import com.fintech.algotrading.instrument.Security;
import com.fintech.algotrading.orders.CancelOrder;
import com.fintech.algotrading.orders.ModifyOrder;
import com.fintech.algotrading.orders.Order;

public class Orderbook {

    private final Security instrument;
    private final Set<Limit> askLimits = new TreeSet<>(Comparator.comparing(Limit::getPrice));
    private final Set<Limit> bidLimits = new TreeSet<>(Comparator.comparing(Limit::getPrice).reversed());
    private final Map<Long, OrderbookEntry> orders = new HashMap<>();

    // Constructor
    public Orderbook(Security instrument) {
        this.instrument = instrument;
    }

    // Getter for the number of orders
    public int getCount() {
        return orders.size();
    }

    // Add an order to the order book
    public void addOrder(Order order) {
        Limit baseLimit = new Limit(order.getPrice());
        addOrderToLimit(order, baseLimit, order.isBuySide() ? bidLimits : askLimits, orders);
    }

    // Static method to add an order to a limit level
    private static void addOrderToLimit(Order order, Limit baseLimit,
                                        Set<Limit> limitLevels, Map<Long, OrderbookEntry> internalBook) {
        Optional<Limit> existingLimit = limitLevels.stream().filter(l -> l.equals(baseLimit)).findFirst();
        Limit limit = existingLimit.orElse(baseLimit);

        if (!existingLimit.isPresent()) {
            limitLevels.add(baseLimit);
        }

        OrderbookEntry orderbookEntry = new OrderbookEntry();
        if (limit.getHead() == null) {
            limit.setHead(orderbookEntry);
            limit.setTail(orderbookEntry);
        } else {
            OrderbookEntry tailPointer = limit.getTail();
            tailPointer.setNext(orderbookEntry);
            orderbookEntry.setPrevious(tailPointer);
            limit.setTail(orderbookEntry);
        }
        internalBook.put(order.getOrderId(), orderbookEntry);
    }

    // Change an existing order
    public void changeOrder(ModifyOrder modifyOrder) {
        OrderbookEntry orderbookEntry = orders.get(modifyOrder.getOrderId());
        if (orderbookEntry != null) {
            removeOrderFromLimit(modifyOrder.toCancelOrder(), orderbookEntry, orders);
            addOrder(modifyOrder.toNewOrder());
        }
    }

    // Check if an order exists
    public boolean containsOrder(long orderId) {
        return orders.containsKey(orderId);
    }

    // Get all bid orders
    public List<OrderbookEntry> getBidOrders() {
        List<OrderbookEntry> orderbookEntries = new ArrayList<>();
        for (Limit bidLimit : bidLimits) {
            if (bidLimit.isEmpty()) continue;
            OrderbookEntry entry = bidLimit.getHead();
            while (entry != null) {
                orderbookEntries.add(entry);
                entry = entry.getNext();
            }
        }
        return orderbookEntries;
    }

    // Get all ask orders
    public List<OrderbookEntry> getAskOrders() {
        List<OrderbookEntry> orderbookEntries = new ArrayList<>();
        for (Limit askLimit : askLimits) {
            if (askLimit.isEmpty()) continue;
            OrderbookEntry entry = askLimit.getHead();
            while (entry != null) {
                orderbookEntries.add(entry);
                entry = entry.getNext();
            }
        }
        return orderbookEntries;
    }

    // Get the current spread
    public OrderbookSpread getSpread() {
        Limit bestAsk = askLimits.stream().filter(l -> !l.isEmpty()).findFirst().orElse(null);
        Limit bestBid = bidLimits.stream().filter(l -> !l.isEmpty()).findFirst().orElse(null);

        Long bestAskPrice = bestAsk != null ? bestAsk.getPrice() : null;
        Long bestBidPrice = bestBid != null ? bestBid.getPrice() : null;

        return new OrderbookSpread(Optional.ofNullable(bestBidPrice), Optional.ofNullable(bestAskPrice));
    }

    // Remove an order from the order book
    public void removeOrder(CancelOrder cancelOrder) {
        OrderbookEntry orderbookEntry = orders.get(cancelOrder.getOrderId());
        if (orderbookEntry != null) {
            removeOrderFromLimit(cancelOrder, orderbookEntry, orders);
        }
    }

    // Static method to remove an order from a limit level
    private static void removeOrderFromLimit(CancelOrder cancelOrder, OrderbookEntry orderbookEntry,
                                             Map<Long, OrderbookEntry> internalBook) {
        Limit limit = orderbookEntry.getParentLimit();
        if (orderbookEntry.getPrevious() != null && orderbookEntry.getNext() != null) {
            orderbookEntry.getNext().setPrevious(orderbookEntry.getPrevious());
            orderbookEntry.getPrevious().setNext(orderbookEntry.getNext());
        } else if (orderbookEntry.getPrevious() != null) {
            orderbookEntry.getPrevious().setNext(null);
        } else if (orderbookEntry.getNext() != null) {
            orderbookEntry.getNext().setPrevious(null);
        }

        if (limit.getHead() == orderbookEntry && limit.getTail() == orderbookEntry) {
            limit.setHead(null);
            limit.setTail(null);
        } else if (limit.getHead() == orderbookEntry) {
            limit.setHead(orderbookEntry.getNext());
        } else if (limit.getTail() == orderbookEntry) {
            limit.setTail(orderbookEntry.getPrevious());
        }

        internalBook.remove(cancelOrder.getOrderId());
    }
}
